creusot-contracts/src/logic/set.rs

```diff
@@ -9,15 +9,54 @@ use std::ops::Index;
 pub struct Set<T: ?Sized>(std::marker::PhantomData<T>);
 
 impl<T: ?Sized> Set<T> {
+    #[trusted]
+    #[creusot::builtins = "set.Set.empty"]
+    pub const EMPTY: Self = { Set(std::marker::PhantomData) };
+
     #[predicate]
-    #[creusot::builtins = "set.Set.mem"]
+    #[why3::attr = "inline:trivial"]
     pub fn contains(self, e: T) -> bool {
-        pearlite! {absurd}
+        Self::mem(e, self)
     }
 
-    #[predicate]
-    #[creusot::builtins = "set.Set.add"]
+    #[creusot::builtins = "set.Set.mem"]
+    fn mem(e: T, set: Self) -> bool {
+        pearlite! { absurd }
+    }
+
+    #[logic]
+    #[why3::attr = "inline:trivial"]
     pub fn insert(self, e: T) -> Self {
+        Self::add(e, self)
```

Comment:
I'm not such a great fan of having plenty of notions which are aliases from each other: insert/add remove/rem mem/contains. Why do we need these aliases?

Reply:
`mem` is private to the `Set` module, it is not exported as part of the API. It's needed as must match the signature of the `why3` function which is flipped wrt Rust conventions.